Introduce a codec_adapter component

mrajwa opened the pull request:
This patch set introduces a "codec_adapter" component which provides interfaces to communicate with codec libraries.
With this component it is possible to link external processing libraries and use them as a regular SoF component, using SoF known API. The current version fully supports all codecs which conforms to the CADENCE API. So if one is going to use codec based on CADENCE interfaces the integration procedure simply boils down to:

a. Link the library
b. Provide its setup data in the topology in the form of C_CONTROL

On the other hand, for codecs which doesn't conform to CADENCE API there is a need to additionaly implement **codec specific**: 
.init, .prepare. .process. .reset and  .free methods. NOTE! As stressed above these are codec specific methods not the ones dedicated for SoF component - this part is handled by codec_adapter component.  

It is possible that codec adapter will fully support more then CADENC API in the future.

dbaluta commented on src/audio/Kconfig:
```diff
@@ -269,6 +269,13 @@ endmenu # "Downsampling ratios"
+	default n
+	help
+	  Select for use of any HiFi external codec libraries.
+	  This component is sort of an adapter for any external HiFi codec.
```
I don't think we should mention HiFi here. This is just a generic processor and in theory should work with any codec.

mrajwa replied:
Yes you are right, will correct this "help message".

lgirdwood replied:
I guess this is intened for 3rd party binary libraries/codecs since open source components would just be added as source via PRs ? If so , we should state this in the help.

mrajwa replied:
yes this component is sort of an adapter between SOF components and 3party libraries as you said.

dbaluta commented on src/audio/generic_processor/generic_processor.c:
```diff
@@ -0,0 +1,94 @@
+	int ret;
+	struct comp_dev *dev = NULL;
+	struct comp_data *cd = NULL;
+
```
No need to initialize these two variables with NULL.

mrajwa replied:
These are auto variables, if we fail at the beginning of.new() and go to error handling code section the `rfree` will try to reclaim this memory. So the NULL assignment makes sure we don't call `rfree` if this particular object has not been allocated.

paulstelian97 replied:
I think for dev it's still unnecessary since there is no branch between these initializations and the actual allocation. But yeah, better safe I guess.

mrajwa replied:
@plbossart, and this is exactly what this code does.

plbossart replied:
no, you need to have different labels and no conditional checks.

mrajwa replied:
this will make code less readable I think, why are you afraid of conditional checks in error handling procedure? This is not sth which is called every milisecods so we shouldn't be concerned about performance at all...if that's what you care about

plbossart replied:
that's how driver code is written, it helps make sure the resources are released in the right order. 
Testing for NULL variables before freeing them is not what we do.

dbaluta commented on src/include/sof/audio/generic_processor/generic_processor.h:
```diff
@@ -15,9 +15,19 @@ enum gp_state {
 struct comp_data {
 	enum gp_state state;
+	struct generic_processor_config gp_config;
+
```
Here we should go either with ``gp_`` prefix or ``generic_processor``. So, I think the structure should be named ``struct gp_config`` and then declare config member inside comp_data.

mrajwa replied:
Agree.

dbaluta commented on src/include/sof/audio/generic_processor/hifi_codec/codec.h:
```diff
@@ -0,0 +1,29 @@
+struct hifi_codec_data {
+	struct hifi_codec_config s_cfg;
+	struct hifi_codec_config r_cfg;
+};
```
What does s_cfg / r_cfg means? At least we should add some comments unless we find a better name.

mrajwa replied:
s_cfg and r_cfg mean setup_config and runtime_config respectively.

paulstelian97 replied:
@mrajwa Use the full names, it won't break any line lengths (especially considering limit is now 100)

dbaluta commented on src/include/sof/audio/generic_processor/hifi_codec/xa_type_def.h:
```diff
@@ -0,0 +1,100 @@
+/*
```
I think we should introduce Cadence provided interface files at the beginning in a separate file, so that we can have all of them in one commit. What do you think?

mrajwa replied:
hmm why would we do so? This would create a bit *dummy* commits without any functionality behind them. I think the idea is to make a clear distinction between Cadence provided code and ours, I sort of buy this idea but in the same time I don't see it as a necessary step.

dbaluta commented:
@mrajwa thanks for sending this! My initial thought about naming this component was "PROCESS" or "PROCESSING". I think your name Generic Processor also works.

mrajwa commented:
@dbaluta, thanks for the review. Once we agree on the naming convention and general framework shape I will continue upstreaming. @lgirdwood @plbossart any comments from you? :)

lgirdwood commented on src/audio/generic_processor/generic_processor.c:
```diff
@@ -0,0 +1,94 @@
+static const struct comp_driver comp_generic_processor;
+
+//TODO: provide proper uuid below
+/* d8218443-5ff3-4a4c-b388-6cfe07b956aa */
```
Is this file meant to be a template to be used by other processing components ? If so, we should state more details of how to use this file in the top level comments.

mrajwa replied:
@lgirdwood I will add some more comments there sure, nevertheless we need a dedicated page in sof-docs for this component

lgirdwood commented on src/include/ipc/topology.h:
```diff
@@ -47,6 +47,7 @@ enum sof_comp_type {
 	SOF_COMP_ASRC,		/**< Asynchronous sample rate converter */
 	SOF_COMP_DCBLOCK,
 	SOF_COMP_SMART_AMP,		/**< smart amplifier component */
+	SOF_COMP_GENERIC_PROCESSOR,	/** generic processor component */
```
@keyonjie dont we have this generic type upstream now ?

mrajwa replied:
@plbossart, if no new types then what SOF_COMP_* should this generic_processor be?

lgirdwood commented on src/include/sof/audio/generic_processor/generic_processor.h:
```diff
@@ -0,0 +1,23 @@
+#define __SOF_AUDIO_GENERIC_PROCESSOR__
+
+enum gp_state {
+	PP_STATE_DISABLED = 0,
```
Best to have some doxygen comments here.

lgirdwood commented on src/include/sof/audio/generic_processor/hifi_codec/public.h:
```diff
@@ -0,0 +1,26 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2019 Intel Corporatiogit an. All rights reserved.
```
2020

lgirdwood replied:
Btw, is this the header to the external binary. If so we should state here that any types/values are immutable.

paulstelian97 replied:
Further, typo (a "git a" somehow made its way in the middle of the line)

dbaluta replied:
I would try to find a fancy name instead of public.h. Maybe api.h or codec_interface.h?

plbossart replied:
Intel Corporatiogit an.

?

lgirdwood reviewed (commented):
I think this is a big step forward, but it does seem we have generic parts and cadence parts here. I think at the high level we should have
```src/audio/cadence``` and ```src/include/audio/cadence``` for all cadence specific files
```src/audio/external``` and ```src/include/audio/external``` for all generic "wrapping files"
This way it's easy to visualise vendor specific from the generic external wrapper.

jajanusz commented on src/audio/generic_processor/generic_processor.c:
```diff
@@ -0,0 +1,140 @@
+	}
+
+	/* Init processing codec */
+	ret = hifi_codec_init(dev, cd->gp_config.codec_id);
```
Why is it bound to some specific codec? (hifi here)
I would imagine processor component to define some generic api that codecs have to implement (and register in similar way to DECLARE_MODULE).

juimonen commented:
@mrajwa sorry for being late for this... but now when we have the uuids, I don't see the "generic component" is a good name as I don't see we have "generic components". It is more like "external binary component interface for xyz": so you see I can't come up with good name, but "generic" is misleading IMHO.

mrajwa commented:
@juimonen, its not "generic_component" but "generic_processor" I am open for a better name still, how about `codec_process`?

mrajwa commented on src/audio/generic_processor/generic_processor.c:
```diff
@@ -0,0 +1,94 @@
+// Author: Marcin Rajwa <marcin.rajwa@linux.intel.com>
+
+/*
+ * A generic processor component.
```
Hmm why do you think it is misleding? I don't like the first two proposal from you because they are not appropriate noun. On the other hand generic_codec or even codec_processor would be fine for me.

plbossart replied:
a processor hints at a hardware implementation, at least in English and French.

this really looks like a wrapper for encoder/decoder libraries. Is there any plan to use this for PCM processing?

mrajwa replied:
Yes PCM too.

plbossart replied:
ok, so it cannot be using 'codec' 
But it's not generic either, this is only for Cadence libraries, right? so what about cadence_adapter?

mrajwa replied:
Not only for Cadence. Also there is a lot of confusion, for example the word `codec` suggest audio stream encoding/decoding functionality however the Cadence codecs can also work with PCM streams. Therefore I feel like `codecq is fine here. More so, they say their API is "common for HiFi audio codec libraries" so we have another keyword "hifi" used here which has many other connotations.

plbossart replied:
@mrajwa please elaborate: who would be using this API? I understand it's use is mainly/only for interfacing with closed-source libraries from Cadence. What others are you referring to?

plbossart commented:
> @juimonen, its not "generic_component" but "generic_processor" I am open for a better name still, how about `codec_processor`?

codec_wrapper?

dbaluta commented:
> @juimonen, its not "generic_component" but "generic_processor" I am open for a better name still, how about `codec_processor`?



> codec_wrapper?


+1 for codec_wrapper

mrajwa commented:
> > @juimonen, its not "generic_component" but "generic_processor" I am open for a better name still, how about `codec_processor`?
> 
> > codec_wrapper?
> 
> +1 for codec_wrapper

I am OK with codec_wrapper as well as codec_adapter. Anybody with some other naming ideas? :) 

PS: personally I think codec_adapter sounds better

juimonen commented:
huh, this is getting difficult :) I don't also like "codec" because I really understand it to be some sort of coder like lc3 etc. However as I understand this can be what ever audio processing (?). OTH I like the adapter part. So what about something like "binary_adapter"? the meaning I'd like the name to carry is "an adapter for external closed source audio processors", although it is difficult to do that in 2 words...

mmaka1 commented on src/audio/generic_processor/CMakeLists.txt:
```diff
@@ -1,3 +1,3 @@
 # SPDX-License-Identifier: BSD-3-Clause
 
-add_local_sources(sof generic_processor.c)
+add_local_sources(sof generic_processor.c hifi_codec/codec.c)
```
hifi_codec name suggests specialization for Hifi co-processor while this layer supports all Cadence Codec API modules and there is no direct reference to any Hifi code.

mmaka1 commented on src/audio/generic_processor/hifi_codec/codec.c:
```diff
@@ -65,3 +65,28 @@ int hifi_codec_load_config(struct comp_dev *dev, void *cfg, size_t size,
+
+	/* select API */
+	if (codec_id >
+	    sizeof(hifi_codec_lib) / sizeof(struct hifi_processing_codec)) {
```
ARRAY_SIZE()

mmaka1 commented on src/audio/generic_processor/hifi_codec/codec.c:
```diff
@@ -65,3 +65,28 @@ int hifi_codec_load_config(struct comp_dev *dev, void *cfg, size_t size,
+	comp_info(dev, "hifi_codec_init() start");
+
+	/* select API */
+	if (codec_id >
```
Is a single uuid shared by all codecs and there is another codec_id? So all instances share same uuid for logging etc?

mrajwa replied:
Yes UUID is single - only codec_adapter gets its own UUID. If you want to create a pipeline with two codecs then you add two codec_adapter components each working with different codec.

ktrzcinx commented on src/audio/generic_processor/generic_processor.c:
```diff
@@ -0,0 +1,140 @@
+	cfg = (struct generic_processor_config *)ipc_generic_processor->data;
+	bs = ipc_generic_processor->size;
+	if (bs) {
+		if (bs < sizeof(struct generic_processor_config)) {
+			comp_info(dev, "generic_processor_new() error: wrong size of setup config");
+			goto err;
+		}
```
Wrong trace function, btw. `ERROR` prefix is added to each `comp_err`/`tr_err` automatically, so there's no need to add them manually in message content.

mrajwa commented:
@lgirdwood @plbossart @dbaluta This patch has just been updated to align with tgl-010-drop-stable.

dbaluta commented on src/audio/codec_adapter/codec_adapter.c:
```diff
@@ -230,6 +230,90 @@ static int codec_adapter_params(struct comp_dev *dev,
 
+static void codec_adapter_copy_to_lib(const struct audio_stream *source,
+				      void *lib_buff, size_t bytes)
+{
```
These two functions can be made generic and moved to audio_stream.h. See: https://github.com/thesofproject/sof/pull/3436/commits/f50fed72fbc8570a032945f75122efcb66297026

dbaluta replied:
@mrajwa is it OK with you to move this functions to audio_stream.h and rename the functions.
These functions should be generic and do not have anything to do with the current patch:

``` codec_adapter_copy_to_lib -> audio_stream_copy_to_buf```

lgirdwood replied:
@dbaluta @mrajwa any update here ?

dbaluta commented on src/audio/codec_adapter/codec_adapter.c:
```diff
@@ -230,6 +230,90 @@ static int codec_adapter_params(struct comp_dev *dev,
+		codec->cpd.avail = codec_buff_size;
+		//TODO: call codec processing function here
+		codec_adapter_copy_from_lib_to_sink(codec->cpd.out_buff,
+						    &sink->stream,
```
The problem with processing function here is that for some decoders e.g mp3, the input data might not be enough and the library asks for more data. 
I'm not sure what's the best direction here:
- when the processing library returns NOT_ENOUGH_DATA we should skip the pipeline walk and allow for more data to arrive?
- another problem is for the output buffer. What happens if processing library cannot copy data to the output buffer?

lgirdwood replied:
> * when the processing library returns NOT_ENOUGH_DATA we should skip the pipeline walk and allow for more data to arrive?

We should just skip this component and continue walking. This way nothing else downstream should starve providing.....

> * another problem is for the output buffer. What happens if processing library cannot copy data to the output buffer?

the input/output buffers here are large enough so that it never starves downstream (i.e. it can cope with variable output from processing to largest output chunks) and never overruns upstream (large enough to store largest decoder block size + overhead).

dbaluta commented on tools/topology/sof/pipe-processing-playback.m4.c:
```diff
@@ -0,0 +1,60 @@
+# Low Latency Passthrough Pipeline and PCM
+#
+# Pipeline Endpoints for connection are :-
```
Commit message should also add BUF0 and BUF1 buffers between components. This would make the commit message more clear.

mrajwa replied:
added.

mrajwa commented on CMakeLists.txt:
```diff
@@ -181,6 +181,10 @@ target_link_libraries(sof PRIVATE sof_options)
 target_link_libraries(sof PRIVATE sof_ld_flags)
 target_link_libraries(sof PRIVATE sof_static_libraries)
+# link codec library for codec_adapter to use
+if(CONFIG_COMP_CODEC_ADAPTER)
```
I made a mistake here, should be **CONFIG_CADENCE_CODEC** of course. Will correct...

paulstelian97 commented on src/audio/codec_adapter/codec_adapter.c:
```diff
@@ -91,6 +91,12 @@ static struct comp_dev *codec_adapter_new(const struct comp_driver *drv,
 
+static inline int validate_setup_config(struct ca_config *cfg)
+{
+	/* TODO: validate codec_adapter setup parameters */
```
Can't this end up being pluggable (specific to each codec)?

mrajwa replied:
This is to validate the config of SoF component codec_adapter. For codecs there is none validation function at the moment, but sure I will add one :)

paulstelian97 commented on src/audio/codec_adapter/codec/generic.c:
```diff
@@ -172,6 +172,47 @@ void *codec_allocate_memory(struct comp_dev *dev, uint32_t size,
 	return ptr;
 }
 
+int codec_free_memory(struct comp_dev *dev, void *ptr)
```
So we're keeping a linked list of allocations so we can clean up just in case the codec requests a stupid deallocation or fails to clean up everything?

mrajwa replied:
Yes, codecs may not keep references of memory you gave them and/or don't provide API calls to deallocate that memory. Hence we let codec allocate memory using our interface so we keep eyes on it.

paulstelian97 commented on src/audio/Kconfig:
```diff
@@ -324,6 +324,16 @@ config COMP_CODEC_ADAPTER
+	bool "Cadence codec"
+	default n
+	help
+		Select for codecs which conforms to the Cacence API.
```
Typo.

mrajwa replied:
thanks, fixed

paulstelian97 commented on src/audio/codec_adapter/codec/generic.c:
```diff
@@ -218,3 +218,36 @@ static int validate_config(struct codec_config *cfg)
+
+	ret = codec->call->prepare(dev);
+	if (ret) {
+		comp_err(dev, "codec_prepare() error %d: codec specific prepare failed, codec_id %x",
```
Might want to add prefix 0x to be less confusing.

paulstelian97 commented on src/audio/codec_adapter/codec/cadence.c:
```diff
@@ -63,6 +63,50 @@ int cadence_codec_init(struct comp_dev *dev)
+		API_CALL(cd, XA_API_CMD_SET_CONFIG_PARAM, param->id,
+			 param->data, ret);
+		if (ret != LIB_NO_ERROR) {
+			comp_err(dev, "apply_config() error %x: failed to applay parameter %d value %d",
```
Typo. apply

paulstelian97 commented on src/audio/codec_adapter/codec/cadence.c:
```diff
@@ -124,7 +226,7 @@ int cadence_codec_prepare(struct comp_dev *dev)
 
 	if (!codec->s_cfg.avail && !codec->s_cfg.size) {
-		comp_err(dev, "cadence_codec_prepare() no setup configuration available!");
+		comp_err(dev, "cadnece_codec_prepare() no setup configuration available!");
```
This should be corrected in the original commit. Please rebase.

paulstelian97 commented on src/audio/codec_adapter/codec/cadence.c:
```diff
@@ -278,6 +278,14 @@ int cadence_codec_prepare(struct comp_dev *dev)
+	if (ret != LIB_NO_ERROR) {
+		comp_err(dev, "cadence_codec_prepare() error %x: failed to initialize codec",
+			 ret);
+		codec_free_memory(dev, cd->mem_tabs);
```
I feel like you're calling codec_free_memory quite a bit, you might want to add a second err2: label?

mrajwa replied:
Yes I agree

dbaluta replied:
@mrajwa Looks like this comment is still not addressed. Sorry for us being picky, this is how a reviewer should be :).

paulstelian97 commented on src/audio/codec_adapter/codec_adapter.c:
```diff
@@ -220,12 +220,21 @@ static int codec_adapter_prepare(struct comp_dev *dev)
+	comp_dbg(dev, "codec_adapter_params(): codec_adapter doesn't support .params() method.");
+
+	return 0;
+}
```
I don't like this commit that adds no functionality. The method is optional, so you might instead want to document that you want to support it in the future, rather than add a dummy implementation. params() will always be called on it if it is made available.

paulstelian97 commented on src/audio/codec_adapter/codec_adapter.c:
```diff
@@ -530,6 +530,19 @@ static int codec_adapter_trigger(struct comp_dev *dev, int cmd)
+	comp_cl_info(&comp_codec_adapter, "codec_adapter_reset(): done");
+
+	return comp_set_state(dev, COMP_TRIGGER_RESET);
+}
```
I feel this implementation is incomplete; doesn't reset also need to do something to the actual codec?

paulstelian97 reviewed (commented):
All in all, I like the general shape of this, but there's obviously a lot of work left on the table. Partial approve. I'd like to also know that this is tested.

mrajwa commented:
@paulstelian97, thanks for the review. What do you mean by "there's obviously a lot of work left on the table" - what would you like to add there?

mrajwa commented on src/audio/generic_processor/generic_processor.c:
```diff
@@ -0,0 +1,94 @@
+
+static const struct comp_driver comp_generic_processor;
+
+//TODO: provide proper uuid below
```
No no, this is not a message for integrators, the UUID for codec adapter will always be the same. The comment was for me to update the temporary UUID I placed.

paulstelian97 commented:
Well, besides the params method which currently has no implementation (it's not required), I also don't like a pattern I see in your commits: commit that introduces a generic function, commit(s) that introduce the specific implementation(s) of the function, then calling the specific variant from the generic variant. Why not put the specific implementations first and then the generic variant is added as complete in the first place?

Plus the typos, and otherwise... does it work?

mrajwa commented:
@paulstelian97, the params method is not needed as we send them using set_binary_data interface. As for the pattern of the commits, well this is how I was developing the code, for example- I started with high level .prepare() method in codec_adapter then I added the genric part of it (shared by all codecs) then I implemented specific method for CADENCE codecs and then stich it all together. 

> Plus the typos, and otherwise... does it work?

Yes it does

paulstelian97 commented:
> As for the pattern of the commits, well this is how I was developing the code, for example- I started with high level .prepare() method in codec_adapter then I added the genric part of it (shared by all codecs) then I implemented specific method for CADENCE codecs and then stitch it all together.

Okay but that does get confusing when reviewing at a commit-by-commit basis, and also you get more commits than necessary. Or at least the fact that the functions with basically no implementation are also done in separate commits instead of one shared commit. Guess I'll wait for a consensus instead of by myself blocking because of this.

lgirdwood commented on src/audio/Kconfig:
```diff
@@ -300,6 +300,16 @@ config COMP_TDFB
+	  be statically linked to the SoF FW binary image and the codec details filled in
+	  the codec_lib[] object. It is possible to link several different codecs and use
+	  them in parallel.
+
```
Where do we specify path to codec_lib ?

mrajwa replied:
Yes good question, I think I will change that description to be a bit more generic.. I think we don't want to provide a complete integration guide here but just highlight few key aspect of codec_adapter usage.

mrajwa commented:
@lgirdwood @plbossart @dbaluta @paulstelian97 I have just updated this PR, several typos were corrected, commit message of last commit extended. Can we proceed to merge? PR as big as this quickly becomes prone to conflicts.

lgirdwood commented on src/include/sof/audio/codec_adapter/codec_adapter.h:
```diff
@@ -0,0 +1,13 @@
+#ifndef __SOF_AUDIO_CODEC_ADAPTER__
+#define __SOF_AUDIO_CODEC_ADAPTER__
+
+#include <sof/audio/codec_adapter/codec/generic.h>
```
What' the header used for  - seems to be some indirection only ?

mrajwa replied:
This header is used by `codec_adapter.h`, `generic.c` and cadence.c it provides "generic" data structures and interfaces shared beetween codec_adapter SoF component and specific codec like CADENCE ones

lgirdwood replied:
ok, but I need to see the big picture. i.e. squash all commit for this new file and the others.

lgirdwood reviewed (changes requested):
@mrajwa can you squash all the changes for any new files into single patches. This make it easier to review for everyone. Changes for existing files are fine as they are now.

mrajwa commented:
@lgirdwood squashing is problematic since it will make potential bisecting very difficult also plenty of commit messages would have to modified. Personally I found github keyboard shortcuts `n` (next) & `p` (previous) to be very useful when you navigate thru patches.

lgirdwood commented:
> @lgirdwood squashing is problematic since it will make potential bisecting very difficult also plenty of commit messages would have to modified. Personally I found github keyboard shortcuts `n` (next) & `p` (previous) to be very useful when navigating thru patches.

Bisect is atomic for new files (with the Makefile/Kconfig changes always added after all new files). The reason that new files are added in a single patch is that it allows the reviewer to see the complete file in one shot, otherwise reviewers have to remember the context from the previous 28 patches (and this is the same way things are done in Linux - for the same reasons). It also lowers the burden for the author too.

mrajwa commented:
Hmm I doubt it will help to grasp the flow. Let's take a .prepare() method as an example here `CODEC_ADAPTER` (a regular SoF component) implements .prepare() method in `codec_adapter.c` file - it basically checks states, assign sinks & sources etc. Then it calls `.prepare()` method in `generic.c` which checks which particular code is attached to this adapter and calls its .prepare method as well. So just to implement complete prepare you need to review 3 different files. Now if I squash all functions like .new, .prepare, .copy .reset etc into one commit then the review IMHO will be even more difficult because instead of looking only on XX new lines of code you would have to dig into XXX lines of code, the commit description will be very vague as well..
Sure I can do this if you want, but personally I would stick to small/per feature commits.
Also if you want to have a whole code at a glance you can go to https://github.com/thesofproject/sof/pull/3361/files

lgirdwood commented:
> Sure I can do this if you want, but personally I would stick to small/per feature commits.

Please do, otherwise I cant easily follow this alongside the other unrelated PRs. I need to see the big picture alongside the commit message.

mrajwa commented:
> > Sure I can do this if you want, but personally I would stick to small/per feature commits.
> 
> Please do, otherwise I cant easily follow this alongside the other unrelated PRs. I need to see the big picture alongside the commit message.

I have just squashed several commits which together form a complete function like .prepare() method. Let me know if you can follow it now. I am also thinking about a review meeting - maybe we can organize it and went thru all these patches?

dbaluta commented on src/audio/codec_adapter/codec_adapter.c:
```diff
@@ -71,12 +72,91 @@ static struct comp_dev *codec_adapter_new(const struct comp_driver *drv,
+ * for the initialization of codec_adapter and follows struct ca_config.
+ * Second contains codec specific data needed to setup the codec itself.
+ * The leter is send in a TLV format organized by struct codec_param.
+ *
```
I assume you want instead of ``The leter is send`` -> ``The latter is sent``? Just to avoid confusion.

mrajwa replied:
yep, thanks for spotting.

dbaluta commented on src/include/sof/audio/codec_adapter/codec/generic.h:
```diff
@@ -55,11 +55,19 @@ struct codec_config {
+	struct codec_memory *prev;
+	struct codec_memory *next;
+};
+
```
should we use list API here? (see list.h?)

paulstelian97 replied:
Eh, in this particular case a bespoke list might be acceptable. I'd say leave it up to him.

dbaluta replied:
@paulstelian97 Ok, no strong preference on this.

mrajwa replied:
We can use the `list.h` API.

dbaluta commented on src/audio/codec_adapter/codec/cadence.c:
```diff
@@ -0,0 +1,36 @@
+#include <sof/audio/codec_adapter/codec/cadence.h>
+
+int cadence_codec_init(struct comp_dev *dev)
+{
```
In the Linux kernel I noticed that most of the functions written for Cadence code use cdns instead of Cadence. Maybe this is to have shorter lines?

So, I think it might worth having a look at this. Instead of cadence_codec_init -> cdns_codec_init.

paulstelian97 replied:
When the Cadence driver was made the line length limit was still 80, so there's that.

dbaluta replied:
@paulstelian97 OK, no hard feelings on this :). We can use cadence instead of cdns :)

dbaluta commented on src/include/sof/audio/codec_adapter/codec/cadence.h:
```diff
@@ -14,6 +14,20 @@
+			  (sub_cmd), \
+			  (value)); \
+	} while (0)
+
```
I don't think we need extra API macro, also the API call can be placed on one line and also API_CALL can be named CODEC_API_CALL.
So, the code should look like this:

```
extern xa_codec_func_t cdns_api_func;

#define CODEC_API_CALL(cd, cmd, sub_cmd, value, ret) \
	do { \
		ret = cdns_api_func((cd)->self, (cmd),   (sub_cmd), (value)); 
	} while (0) 

```

mrajwa replied:
No need to preceed it with `CADENCE` or `CDNS` either, this will be called from `cadence.c` only.

dbaluta replied:
@mrajwa all right.

dbaluta commented on src/audio/codec_adapter/codec/generic.c:
```diff
@@ -251,3 +251,30 @@ int codec_prepare(struct comp_dev *dev)
+	}
+
+	ret = codec->call->process(dev);
+	if (ret) {
```
I feel that instead of ``codec->call->process`` we should have ``codec->ops->process``

dbaluta commented on src/audio/codec_adapter/codec_adapter.c:
```diff
@@ -324,18 +324,75 @@ static int codec_adapter_copy(struct comp_dev *dev)
+
+static int codec_adapter_ctrl_set_data(struct comp_dev *dev,
+				       struct sof_ipc_ctrl_data *cdata)
+{
```
All static functions in this file should have the same prefix. Either ``codec_adapter`` or ``ca`` but not both. I would prefer ``ca`` as it is shorter, but any works for me.

dbaluta commented on src/audio/codec_adapter/codec_adapter.c:
```diff
@@ -382,6 +382,43 @@ static int ca_set_params(struct comp_dev *dev, struct sof_ipc_ctrl_data *cdata,
+			} else {
+				comp_dbg(dev, "ca_set_params() load of setup config done.");
+			}
+
```
No need for braces here.
```
if (ret)
  do_a();
else
  do_b();

```

dbaluta commented:
@mrajwa @lgirdwood @paulstelian97 I have added more comments. This is definitely going into the right direction.

- @mrajwa I understand your point that is hard to do rebasing. We all want to merge this as soon as possible. I think we can do this be the end of the week after addressing all comments / concerns.

dbaluta commented on src/include/sof/audio/codec_adapter/codec/generic.h:
```diff
@@ -44,8 +56,10 @@ struct codec_config {
 	struct codec_config s_cfg; /**< setup config */
 	struct codec_config r_cfg; /**< runtime config */
+	struct codec_interface *call;
 };
```
The codec_interface field name should be called ops instead of call. As we traditionally call this kind of operations in SOF.

mrajwa replied:
Just a matter of taste I am fine with both proposals.

dbaluta replied:
@mrajwa if you don't mind please use ``ops``.

mrajwa replied:
ok, done.

paulstelian97 commented on src/audio/codec_adapter/codec/generic.c:
```diff
@@ -0,0 +1,79 @@
+	}
+
+	/* Config loaded, mark it as valid */
+	dst->size = size;
```
Might want to move this size assignment next to the allocation, otherwise we might get memory management surprises later on.

mrajwa replied:
I don't get this proposal. You start with `size = 0` and if allocation + config load + validation all went well you set `size` as a mark of valid config. Otherwise you just free the allocated memory and `size` remains 0.

paulstelian97 reviewed (commented):
Looks better but I have one general question: How will codec_adapter interact with multiple DSP cores? I see no protection or attempt at atomicity in this sense.

mrajwa commented:
> Looks better but I have one general question: How will codec_adapter interact with multiple DSP cores? I see no protection or attempt at atomicity in this sense.

You just schedule it on a core of choice and that's all,  the .copy() function takes care of invalidation of memory and writebacks.

lgirdwood commented on src/include/sof/audio/codec_adapter/codec/generic.h:
```diff
@@ -0,0 +1,32 @@
+/* Codec generic data types						     */
+/*****************************************************************************/
+enum ca_state {
+	PP_STATE_DISABLED = 0,
```
Can we have doxygen comments for all the new API structs and enums. This code will be used by lots of external folks so it need to be documented.

lgirdwood reviewed (commented):
I'm still seeing 31 commits ? Shouldn't this be less with squashing of new files ?

mrajwa commented:
@lgirdwood it was 50+ commits before, well if you want me to really do it on file basics then I need to manual rework whole patch set...plain squashing won't do it.. working on it now...

mrajwa commented:
@lgirdwood, @dbaluta @paulstelian97 patch set updated according to your suggestions.

dbaluta commented on src/audio/codec_adapter/codec_adapter.c:
```diff
@@ -0,0 +1,515 @@
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		comp_warn(dev, "codec_adapter_cmd() set_data not implemented yet.");
+		ret = codec_adapter_ctrl_set_data(dev, cdata);
```
Remove comp_warn for COMP_CMD_SET_DATA.

lgirdwood replied:
Looks like this is still here.

dbaluta commented on src/audio/codec_adapter/codec_adapter.c:
```diff
@@ -0,0 +1,515 @@
+	rfree(cd);
+	rfree(dev);
+
+	comp_cl_info(&comp_codec_adapter, "codec_adapter_free(): component memory freed");
```
Usually we only add debug calls at the beginning of the functions. Please remove the comp_cl_info at the end.

lgirdwood replied:
ack

lgirdwood replied:
ping ?

dbaluta commented on src/include/sof/audio/codec_adapter/codec/generic.h:
```diff
@@ -0,0 +1,115 @@
+	CODEC_INITIALIZED,
+	CODEC_PREPARED,
+	CODEC_RUNNING
+};
```
Comma afer CODEC_RUNNING, at some point we might add other states.

dbaluta commented on src/audio/codec_adapter/codec/generic.c:
```diff
@@ -0,0 +1,359 @@
+	if (!dst->data) {
+		dst->data = rballoc(0, SOF_MEM_CAPS_RAM, size);
+	} else if (dst->size != size) {
+		rfree(dst->data);
```
what is the purpose of rfree here?

lgirdwood replied:
ack, this could do with more inline comments.

dbaluta commented on src/audio/codec_adapter/codec/generic.c:
```diff
@@ -0,0 +1,359 @@
+static int validate_config(struct codec_config *cfg)
+{
+	//TODO: validation of codec specifig setup config
+	return 0;
```
s/specifig/specific

dbaluta commented:
@mrajwa please also check the checkpatch.pl warnings. All of them should be addressed, except the ones over the Cadence files which we will keep as they are!

lgirdwood commented on src/audio/codec_adapter/codec_adapter.c:
```diff
@@ -0,0 +1,515 @@
+		} else if (!size) {
+			comp_err(dev, "codec_adapter_set_params() error: no configuration size %d",
+				 size);
+			//TODO: return -EINVAL. This is temporary until driver fixes its issue
```
Do we have a driver bug reported for this ? @kv2019i @juimonen fyi.

mrajwa replied:
Not yet, but I am preparing a patch for it.

lgirdwood replied:
any update ?

lgirdwood commented on src/include/sof/audio/codec_adapter/codec/generic.h:
```diff
@@ -0,0 +1,115 @@
+	int (*process)(struct comp_dev *dev);
+	int (*apply_config)(struct comp_dev *dev);
+	int (*reset)(struct comp_dev *dev);
+	int (*free)(struct comp_dev *dev);
```
Can we have some doxygen comment here explaining each member.

paulstelian97 replied:
Typically we put the doxygen comments next to each member instead of using \var. Some devs like myself read the code comments themselves for documentation for each interface, so I'd expect the styling to match the existing style in other interfaces at the very least.

lgirdwood commented on src/include/sof/audio/codec_adapter/codec/generic.h:
```diff
@@ -0,0 +1,115 @@
+/* CODEC configuration types */
+enum codec_cfg_type {
+	CODEC_CFG_SETUP,
+	CODEC_CFG_RUNTIME
```
ditto for doxygen

lgirdwood commented on src/audio/codec_adapter/codec/cadence.c:
```diff
@@ -0,0 +1,362 @@
+		comp_dbg(dev, "cadence_codec_init(): allocated %d bytes for lib object",
+			 obj_size);
+	}
+
```
This needs documented, what is each stage doing ?

lgirdwood commented on src/audio/codec_adapter/codec/cadence.c:
```diff
@@ -0,0 +1,362 @@
+		}
+		data = (char *)data + param->size;
+		size -= param->size;
+	}
```
Loop need commented. What are we doing ?

lgirdwood commented on src/audio/codec_adapter/codec/cadence.c:
```diff
@@ -0,0 +1,362 @@
+		}
+
+		comp_dbg(dev, "init_memory_tables: allocated memory of %d bytes and alignment %d for mem. type %d",
+			 mem_size, mem_alignment, mem_type);
```
Some comments here needed too - one for each API_CALL

lgirdwood commented on src/include/sof/audio/codec_adapter/interfaces.h:
```diff
@@ -11,9 +11,22 @@
-static struct codec_interface interfaces[0];
+static struct codec_interface interfaces[] = {
+	{
+		.id = 0xCADE01, /**< Cadence interface */
```
Is this a version or an identifier for a certain Cadence component ? or does it apply to all Cadence ?

mrajwa replied:
@lgirdwood, it applies for codec which conforms to Cadence API

lgirdwood reviewed (changes requested):
Thanks @mrajwa I can see the big picture now. One thing we do need is more inline comments for this infrastructure explaining what each function or code block does. This is especially important since this is the main API used by external audio processing developers.

mrajwa commented on src/audio/codec_adapter/codec_adapter.c:
```diff
@@ -0,0 +1,515 @@
+	return 0;
+}
+
+static void codec_adapter_copy_to_lib(const struct audio_stream *source,
```
Agree, done.

mrajwa commented on src/audio/codec_adapter/codec_adapter.c:
```diff
@@ -0,0 +1,515 @@
+		memcpy((char *)lib_buff + head_size,
+		       audio_stream_wrap(source,
+					 (char *)source->r_ptr + head_size),
+					 tail_size);
```
This code should not be touched and not necessarily investigated by integrator, but sure I will comment it.

mrajwa commented:
@lgirdwood, @dbaluta I have just pushed another update, thanks for your feedback I hope I have addressed all your suggestions :)

dbaluta commented:
@mrajwa this looks good enough now.

mrajwa commented:
@paulstelian97 I object, I think it is much more readable this way. The comment section before the struct organizes it nice I see no reason to mix code and comments in this case. Also it will make code more difficult to maintain as these comments span in more than one line I will always have to take core of alignment etc..

paulstelian97 commented:
> @paulstelian97 I object, I think it is much more readable this way. The comment section before the struct organizes it nice I see no reason to mix code and comments in this case. Also it will make code more difficult to maintain as these comments span in more than one line you will always have to take care of alignment etc..

If you believe this coding style is better than the one in the existing structures, I invite you to make a pull request to adjust existing structures to this style.

mrajwa commented:
@paulstelian97 I've never seen struct declaration with long comments mixed with code. Sure there are ones where each member has short, few words inlined comment but here we have comments which span for two lines that's why I think the way I did it is better.

paulstelian97 commented:
> @paulstelian97 I've never seen struct declaration with long comments mixed with code. Sure there are ones where each member has short, few words inlined comment but here we have comments which span for two lines that's why I think the way I did it is better.

`src/include/sof/audio/component.h`, `struct comp_ops` is the best example I could find.

mrajwa commented:
@paulstelian97 ok I see your intention now, I have refactored these comments, is it ok for you now?

paulstelian97 commented:
Yeah, looks mostly good. Now I wonder, what's gonna happen if we want _multiple_ Cadence codecs? How does that work?

mrajwa commented:
@mrajwa you create the topology with X number of codec adapters that's all - each one is working with different cadence or other codec

paulstelian97 commented:
> @paulstelian97 you create the topology with X number of codec adapters that's all - each one is working with different cadence or other codec

I mean, I have multiple Cadence codecs, I haven't seen where the specific implementations for each of them comes in. How does it work? Let's just say 15 Cadence codecs, where do I decide which of them I use in each component?

mrajwa commented:
@paulstelian97, I have just made it more clear - please take a look at `src\audio\codec_adapter\codec\cadence.c` there you have a table of API functions for cadence in this case (each codec type will have its own table). Currently there is one, namely `cadence_api_function` with `id = 0x01`. Now in the topology you have a `C_CONTROL` in which you specify codec_id. For cadence the IDs last three most significant bytes are 0xCADE01 - this specify `cadence` interface, the lowest byte `0x01` in this case is an API id. Does it make sense for you?

dbaluta commented:
@mrajwa I understand the way you select codec id. My only concern is that this is not inline with the compress API.
In the compress API the codec ID is selected at runtime. 

I wonder if we can go right now with our approach (statically selecting the codec to be used, this means you need to have different topologies for each codec, we have around 12 codecs :D).

Cc: @plbossart

mrajwa commented:
@dbaluta its easy to add runtime id handling. We can go like this -> If runtime `codec_id ` has been provided then use it, if not use the one from the topology. Does it address the problem?

dbaluta commented:
@mrajwa yes, this looks good.

paulstelian97 reviewed (approved):
Nothing jarring left, this looks good enough from my view.

mrajwa commented:
@dbaluta we first need to agree how this extra parameter will originate from driver side. Do you have some draft already?

dbaluta commented:
@mrajwa no draft right now, I hope to  start Compress API driver implementation next week. We still have some annoying bugs in our SDMA driver.

mrajwa commented:
Fine, so we can return to runtime specific parameters passing later on.
@lgirdwood, can we proceed with merge now?

lgirdwood commented on src/audio/codec_adapter/codec_adapter.c:
```diff
@@ -0,0 +1,528 @@
+						      codec_buff_size);
+		buffer_invalidate(source, codec_buff_size);
+		codec->cpd.avail = codec_buff_size;
+		//TODO: call codec specific processing here
```
Anything todo here prior to merge ?

mrajwa replied:
No, this is added only to the initial commit in the series. Since you wanted it to be split in the file basics I had to add placeholders for the call which will occur later one. If you check the version of codec_adapter.c after the last commit you will see that `TODO` has been replaced with the actual call to codec processing function.

lgirdwood replied:
ok, but for next time it's far easier and acceptable just to add a new file as a single commit (no need to show the development history here).

lgirdwood replied:
oh, this should be C style comment too.

lgirdwood commented on src/audio/codec_adapter/codec_adapter.c:
```diff
@@ -0,0 +1,528 @@
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_ENUM:
+		//TODO
```
?

mrajwa replied:
There is no need for `SOF_CTRL_CMD_ENUM` so I have added a `TODO` here. I can replace it with a error trace as well.

lgirdwood replied:
error is good. Lets everyone know it's not implemented.

lgirdwood commented on src/include/sof/audio/codec_adapter/codec/generic.h:
```diff
@@ -0,0 +1,189 @@
+	/**
+	 * Codec specific prepare procedure, called as part of codec_adapter
+	 * component
+	 * preparation in .prepare()
```
Can be merged in above line. Btw, can we add what calls are optional and mandatory.

mrajwa replied:
Ok so, the waring in `COMP_CMD_SET_DATA` has been removed as this functionality has been implemented. 
Do you have some concerns with `comp_cl_info(&comp_codec_adapter, "codec_adapter_free(): component memory freed");` not sure what you meant by `ack` in this regards? The comment lin `generic.h` has been merged. As for what methods are mandatory - all of them are.

lgirdwood commented on src/audio/codec_adapter/codec/generic.c:
```diff
@@ -0,0 +1,366 @@
+			 codec_id);
+		ret = -EIO;
+		goto out;
+	} else if (!interface->init) {//TODO: verify other interfaces
```
Still TODO ?

mrajwa replied:
Updated.

lgirdwood commented on src/audio/codec_adapter/codec/generic.c:
```diff
@@ -0,0 +1,366 @@
+
+static int validate_config(struct codec_config *cfg)
+{
+	//TODO: validation of codec specific setup config
```
Can we have all the pending TODOs in C style comments.

mrajwa replied:
Done.

lgirdwood commented on src/audio/codec_adapter/codec/generic.c:
```diff
@@ -0,0 +1,366 @@
+
+	comp_dbg(dev, "codec_prepare() start");
+
+	if (cd->codec.state == CODEC_PREPARED)
```
Can we have an inline comment here explaining why we would be in the prepared state already ?

mrajwa replied:
Yes, so in essence after the reset of the pipeline only the runtime config is cleared so there is no need to prepare the codec again. Comment added.

lgirdwood commented on src/audio/codec_adapter/codec/generic.c:
```diff
@@ -0,0 +1,366 @@
+	codec->r_cfg.data = NULL;
+	list_init(&codec->memory.mem_list);
+
+	if (codec->r_cfg.data)
```
Why do we free cfg data here, can we have a comment.

mrajwa replied:
Yes, cfg has been applied during the prepare procedure so we no longer need it. Comment added as well.

lgirdwood commented on src/audio/codec_adapter/codec/generic.c:
```diff
@@ -0,0 +1,366 @@
+
+	comp_dbg(dev, "codec_process() start");
+
+	if (cd->codec.state < CODEC_PREPARED) {
```
Can we also have a short state transition diagram where we define the states. This will make the rules easier to follow for everyone.

mrajwa replied:
In the code or better in the sof-docs?

lgirdwood replied:
In the code via doxygen is best, that way sof-docs will get it for free.

lgirdwood commented on src/audio/codec_adapter/codec/generic.c:
```diff
@@ -0,0 +1,366 @@
+	if (cd->codec.state < CODEC_PREPARED)
+		return -EPERM;
+
+	if (cd->codec.state < CODEC_PREPARED) {
```
We are checking the same state twice here (see above).

mrajwa replied:
Right, fixed

lgirdwood commented on src/audio/codec_adapter/codec/generic.c:
```diff
@@ -0,0 +1,366 @@
+	rfree(codec->r_cfg.data);
+	codec->r_cfg.data = NULL;
+
+	comp_dbg(dev, "codec_apply_config() end");
```
Dont need the end debug

mrajwa replied:
Are you sure? If there is some problem with that (investigation of some BUG) its good to have both starting and ending logs - if the ending one is missing you know sth wrong happened in between. Do you agree?

paulstelian97 replied:
Insert ending logs whenever you are actually debugging. Normal runtime it's too much spam.

mrajwa replied:
@paulstelian97, hence `comp_dbg()` - you enable when needed.

paulstelian97 replied:
Still atypical but not enough for me to block.

lgirdwood replied:
The only place I would expect to see ```dbg("func enter");``` and ```dbg("func leave");``` is around any API call into 3rd party binary code. i.e. untrusted code.

mrajwa replied:
@lgirdwood correct and this is exactly what happens here -> `ret = codec->ops->apply_config(dev);` does call 3rd party binary API to apply that config by **`API_CALL(cd, XA_API_CMD_SET_CONFIG_PARAM, param->id, param->data, ret);`**

Does it convince you or let's just remove that end comment?

lgirdwood replied:
I would put a comment around any applicable start/end debug stating this is around an "untrusted binary code" so it's clear for anyone debugging.

lgirdwood commented on src/audio/codec_adapter/codec/generic.c:
```diff
@@ -0,0 +1,366 @@
+		goto out;
+	}
+
+	cd->codec.r_cfg.avail = false;
```
Can we have an inline comment here explaining why we are freeing this config data.

mrajwa replied:
Added.

lgirdwood commented on src/audio/codec_adapter/codec/generic.c:
```diff
@@ -0,0 +1,366 @@
+	codec->r_cfg.size = 0;
+	rfree(codec->r_cfg.data);
+
+	codec->state = CODEC_PREPARED;
```
Are we really prepared after reset ? Shouldn't prepare need to be called ?

mrajwa replied:
Yes the idea is to just reset codec to its initial state - reset runtime configuration like volume up/down, certain frequencies damping etc, but the memory allocated for the codec should remain intact.

lgirdwood replied:
ok, best to have a comment here describing.

lgirdwood reviewed (changes requested):
The only big thing is the state machine, everything else is minor.

lgirdwood reviewed (changes requested):
I think we are almost good to merge, just a few minor opens still to address.

mrajwa commented:
@lgirdwood, getting back to this PR, I have just updated it with additional comments you mentioned. Are we good to merge it?

lgirdwood reviewed (approved):
@dbaluta @paulstelian97 good for you now ?

paulstelian97 reviewed (approved):
I think this looks good.

paulstelian97 commented:
@dbaluta is on PTO, coming back on Thursday so don't expect him to review until then.

lgirdwood commented:
@paulstelian97 ok, lets wait until Thursday.

mrajwa commented:
@lgirdwood, I think there is no need to wait for @dbaluta  approvement as he already approved this PR :)

dbaluta commented:
@mrajwa Yes, already approved. This PR looks good to me. Very likely that we will more patches on this but is good to merge now.

mrajwa commented:
@dbaluta that's true of course.
@lrgirdwo, ok to merge now?
